### PR TITLE
Millisecond and microsecond support in date_diff / age functions

### LINKED ISSUE #46160: millsecond and microsecond support in date_diff and age functions
Currently the date_diff and age functions go to a maximum precision of seconds.
It would be good if they also support milli and microseconds too.

### PR DESCRIPTION
**Pull request by kardymonds**
Close https://github.com/ClickHouse/ClickHouse/issues/46160

The task actually consistsed of tasks:

* new units (millisecond/microsecond) for "date_diff() and age()"
* microsecond precision for age(), (date_diff() works with precision = unit)
* bug in `DateTimeTransforms.h` in components.whole for negative fractional value (affect to `date_diff/age` and other functions (like `toYYYYMMDDhhmmss()`) )

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions "date_diff() and age()" now support millisecond/microsecond unit and work with microsecond precision.

### Documentation entry for user-facing changes

- [V ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit d2635bff64464e48c04f1f2675a33a3376adcc96 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51291/d2635bff64464e48c04f1f2675a33a3376adcc96/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Functions/DateTimeTransforms.h [hunk 1]
```diff
@@ -1377,6 +1380,35 @@ struct ToRelativeSecondNumImpl
     using FactorTransform = ZeroTransform;
 };
 
+template <Int64 scale_multiplier>
```
[the remark this answers is not in the document]
Comment (kardymonds) [reply src/Functions/DateTimeTransforms.h:1383]: Added static_assert

File: src/Functions/dateDiff.cpp [hunk 2]
```diff
@@ -373,6 +398,10 @@ class FunctionDateDiff : public IFunction
             impl.template dispatchForColumns<ToRelativeMinuteNumImpl<ResultPrecision::Extended>>(x, y, timezone_x, timezone_y, res->getData());
         else if (unit == "second" || unit == "ss" || unit == "s")
             impl.template dispatchForColumns<ToRelativeSecondNumImpl<ResultPrecision::Extended>>(x, y, timezone_x, timezone_y, res->getData());
+        else if (unit == "millisecond" || unit == "ms")
+            impl.template dispatchForColumns<ToRelativeSubsecondNumImpl<DecimalUtils::scaleMultiplier<DateTime64>(millisecond_scale)>>(x, y, timezone_x, timezone_y, res->getData());
```
[the remark this answers is not in the document]
Comment (kardymonds) [reply src/Functions/dateDiff.cpp]: Ok, used const multiplier  (without scaleMultiplier())

File: docs/zh/sql-reference/functions/date-time-functions.md [hunk 3]
```diff
@@ -625,6 +625,7 @@ SELECT date_add(YEAR, 3, toDate('2018-01-01'));
 │                                    2021-01-01 │
 └───────────────────────────────────────────────┘
 ```
+## age {#age}
```
[the remark this answers is not in the document]
Comment (kardymonds) [reply docs/zh/sql-reference/functions/date-time-functions.md]: Removed

**Review by hanfei1991 — APPROVED** [review]

**Comment by hanfei1991** [comment]
I don't know why I cannot merge master for this PR ... @kardymonds could you update master to re-run the ci so that I can merge it

**Comment by alexey-milovidov** [comment]
This PR added a flaky test that compromises our green CI.
Let's fix it ASAP or revert and fix it then.

**Comment by alexey-milovidov** [comment]
@kardymonds, it's preferred to add a new test for every feature or improvement instead of modifying an old test.

**Comment by rvasin** [comment]
Dmitry (@kardymonds) will try to fix the flaky test on Monday.

**Pull request merged**